src/components/ik-controller.js

```diff
@@ -0,0 +1,234 @@
+    this.iRootToChest = new Matrix4();
+
+    this.headLastVisible = true;
+    this.leftHandLastVisible = true;
+    this.rightHandLastVisible = true;
+
+    this.leftHandRotation = new Matrix4().makeRotationFromEuler(new Euler(-Math.PI / 2, Math.PI / 2, 0));
+    this.rightHandRotation = new Matrix4().makeRotationFromEuler(new Euler(Math.PI / 2, Math.PI / 2, 0));
+  },
+
+  update(oldData) {
+    if (this.data.leftEye !== oldData.leftEye) {
+      this.leftEye = this.el.querySelector(this.data.leftEye);
+    }
+
+    if (this.data.rightEye !== oldData.rightEye) {
+      this.rightEye = this.el.querySelector(this.data.rightEye);
+    }
+
+    if (this.data.middleEye !== oldData.middleEye) {
+      this.middleEye = this.el.querySelector(this.data.middleEye);
+    }
+
+    if (this.data.head !== oldData.head) {
+      this.head = this.el.querySelector(this.data.head);
+    }
+
+    if (this.data.neck !== oldData.neck) {
+      this.neck = this.el.querySelector(this.data.neck);
+    }
+
+    if (this.data.leftHand !== oldData.leftHand) {
+      this.leftHand = this.el.querySelector(this.data.leftHand);
+    }
+
+    if (this.data.rightHand !== oldData.rightHand) {
+      this.rightHand = this.el.querySelector(this.data.rightHand);
+    }
+
+    if (this.data.chest !== oldData.chest) {
+      this.chest = this.el.querySelector(this.data.chest);
+    }
+
+    if (this.data.hips !== oldData.hips) {
+      this.hips = this.el.querySelector(this.data.hips);
+    }
+
+    // Set middleEye's position to be right in the middle of the left and right eyes.
+    const middleEyePosition = this.middleEye.object3D.position;
+    middleEyePosition.addVectors(this.leftEye.object3D.position, this.rightEye.object3D.position);
+    middleEyePosition.divideScalar(2);
+    this.middleEye.object3D.updateMatrix();
+    this.iMiddleEyeToHead.getInverse(this.middleEye.object3D.matrix);
+
+    this.iHipsToHeadVector
+      .addVectors(this.chest.object3D.position, this.neck.object3D.position)
+      .add(this.head.object3D.position)
+      .negate();
+  },
+
+  updateIkRoot(ikRoot) {
+    this.ikRoot = ikRoot;
+  },
+
+  tick(time, dt) {
+    if (!this.ikRoot) {
+      return;
+    }
+
+    const { camera, leftController, rightController } = this.ikRoot;
+    const {
+      hips,
+      head,
+      chest,
+      cameraForward,
+      headTransform,
+      iMiddleEyeToHead,
+      iHipsToHeadVector,
+      flipY,
+      cameraYRotation,
+      cameraYQuaternion,
+      hipsQuaternion,
+      headQuaternion,
+      leftHand,
+      rightHand,
+      rootToChest,
+      iRootToChest
+    } = this;
+
+    // Camera faces the -Z direction. Flip it along the Y axis so that it is +Z.
+    camera.object3D.updateMatrix();
+    cameraForward.multiplyMatrices(camera.object3D.matrix, flipY);
+
+    headTransform.multiplyMatrices(cameraForward, iMiddleEyeToHead);
+    hips.object3D.position.setFromMatrixPosition(headTransform).add(iHipsToHeadVector);
```

Comment:
// Then position the hips such that the head is aligned with headTransform (which positions middleEye in line with the hmd)